homeassistant/components/nws/weather.py

```diff
@@ -0,0 +1,378 @@
+    {
+        vol.Optional(CONF_NAME): cv.string,
+        vol.Optional(CONF_LATITUDE): cv.latitude,
+        vol.Optional(CONF_LONGITUDE): cv.longitude,
+        vol.Optional(CONF_MODE, default="daynight"): vol.In(FORECAST_MODE),
+        vol.Optional(CONF_STATION): cv.string,
+        vol.Required(CONF_API_KEY): cv.string,
+    }
+)
+
+
+def convert_condition(time, weather):
+    """
+    Convert NWS codes to HA condition.
+
+    Choose first condition in CONDITION_CLASSES that exists in weather code.
+    If no match is found, return first condition from NWS
+    """
+    conditions = [w[0] for w in weather]
+    prec_probs = [w[1] or 0 for w in weather]
+
+    # Choose condition with highest priority.
+    cond = next(
+        (
+            key
+            for key, value in CONDITION_CLASSES.items()
+            if any(condition in value for condition in conditions)
+        ),
+        conditions[0],
+    )
+
+    if cond == "clear":
+        if time == "day":
+            return "sunny", max(prec_probs)
+        if time == "night":
+            return "clear-night", max(prec_probs)
+    return cond, max(prec_probs)
+
+
+async def async_setup_platform(hass, config, async_add_entities, discovery_info=None):
+    """Set up the NWS weather platform."""
+
+    latitude = config.get(CONF_LATITUDE, hass.config.latitude)
+    longitude = config.get(CONF_LONGITUDE, hass.config.longitude)
+    station = config.get(CONF_STATION)
+    api_key = config[CONF_API_KEY]
+    mode = config[CONF_MODE]
+
+    if None in (latitude, longitude):
+        _LOGGER.error("Latitude/longitude not set in Home Assistant config")
+        return
+
+    websession = async_get_clientsession(hass)
+    # ID request as being from HA, pynws prepends the api_key in addition
+    api_key_ha = f"{api_key} homeassistant"
+    nws = SimpleNWS(latitude, longitude, api_key_ha, mode, websession)
+
+    _LOGGER.debug("Setting up station: %s", station)
+    try:
+        await nws.set_station(station)
+    except ERRORS as status:
+        _LOGGER.error(
+            "Error getting station list for %s: %s", (latitude, longitude), status
+        )
+        raise PlatformNotReady
+
+    _LOGGER.debug("Station list: %s", nws.stations)
+    _LOGGER.debug(
+        "Initialized for coordinates %s, %s -> station %s",
+        latitude,
+        longitude,
+        nws.station,
+    )
+
+    async_add_entities([NWSWeather(nws, mode, hass.config.units, config)], True)
+
+
+class NWSWeather(WeatherEntity):
+    """Representation of a weather condition."""
+
+    def __init__(self, nws, mode, units, config):
+        """Initialise the platform with a data instance and station name."""
+        self.nws = nws
+        self.station_name = config.get(CONF_NAME, self.nws.station)
+        self.is_metric = units.is_metric
+        self.mode = mode
+
+        self.observation = None
+        self._forecast = None
+
+    @Throttle(MIN_TIME_BETWEEN_UPDATES)
```

Comment:
Should we adjust the default scan interval too, by setting `SCAN_INTERVAL` constant?

Reply:
Do I simply just set the constant?  I see where it is set in the weather component, [here](https://github.com/home-assistant/home-assistant/blob/dev/homeassistant/components/weather/__init__.py#L42-L43) but I'm not sure how to modify it for the platform.

Reply:
Yep. Just define it at the module level.

Reply:
I made SCAN_INTERNAL 15 minutes, and the MIN_TIME_BETWEEN_UPDATES 5 minutes.  I haven't seen any stations that provide data in much more frequent intervals.

Reply:
I have seen cases where the time between two updates is less than 5 minutes - this is typically when wind gusts or visibility are changing rapidly, but it would not be a good idea to poll more frequently.

The normal update cycle for METAR stations is once per hour, plus updates for important changes between the hourly updates.   Unfortunately, these updates often only include the item or items that have changed, so if you only look at the latest update record, you can easily miss changes of other values.  This is why I had my sensor code retrieving a number of records and processing any that were newer than the previously processed update.  Without this, it is quite common to not get good values for some parameters for long periods of time:   unless the poll happens to catch the update with full data.  (Note that non-airport stations typically do not include a raw METAR record)

Reply:
I captured this idea here.  https://github.com/MatthewFlamm/pynws/issues/15